File: Python/gui/PuzzleViewer.py
#!/usr/bin/env python3

from PyQt4 import QtGui
from PyQt4 import QtCore

from core.Field import Field

class PuzzleViewer ( QtGui.QWidget ):
   """Viewer of the puzzle while solving"""
   def __init__ ( self ):
      super ( PuzzleViewer , self ).__init__ ()
      # Create components
      self.__grid_layout = QtGui.QGridLayout ()
      self.__icons = [] # Create empty list, which will contain the rows of icons of the image
      # Add components
      self.setLayout ( self.__grid_layout ) 
      # Configure components
      for i in range ( 0 , 15 ):
         row = []
         for j in range ( 0 , 15 ):
            pixmap = QtGui.QPixmap ( "resources/n.png" )
            label = QtGui.QLabel ()
            label.setPixmap ( pixmap.scaledToHeight ( 40 , QtCore.Qt.SmoothTransformation ) )
            row.append ( label )
            self.__grid_layout.addWidget ( label , i , j )
         self.__icons.append ( row )

   def clean ( self ):
      for i in range ( 0 , 15 ):
         for j in range ( 0 , 15 ):
            pixmap = QtGui.QPixmap ( "resources/n.png" )
            self.__icons[ i ][ j ].setPixmap ( pixmap.scaledToHeight ( 40 , QtCore.Qt.SmoothTransformation ) )
            self.__icons[ i ][ j ].update ()
            
   def update ( self , field ):
      for i in range ( 0 , field.getHeight () ):
         for j in range ( 0 , field.getWidth () ):
            zero= QtGui.QPixmap ( "resources/" + str ( field.getPoint ( j , i ) ) + ".png" )
            self.__icons[ i ][ j ].setPixmap ( zero.scaledToHeight ( 40 , QtCore.Qt.SmoothTransformation ) )
            self.__icons[ i ][ j ].update ()
   
      QtGui.QApplication.processEvents ()
   
   def setSize ( self , new_width , new_height ):
      for i in range ( 0 , new_height ):
         for j in range ( 0 , new_width ):
            pixmap = QtGui.QPixmap ( "resources/0.png" )
            self.__icons[ i ][ j ].setPixmap ( pixmap.scaledToHeight ( 40 , QtCore.Qt.SmoothTransformation ) )
            self.__icons[ i ][ j ].update ()

File: CPP/src/gui/puzzleviewer/puzzleviewer.cc
# include "puzzleviewer.hh" 

Gui::PuzzleViewer::PuzzleViewer ( QWidget* parent )
   : QWidget ( parent )
{
   buildGui ();
}

Gui::PuzzleViewer::~PuzzleViewer ( void )
{
}

void Gui::PuzzleViewer::buildGui ( void )
{
   grid_layout = new QGridLayout ( this );
   setLayout ( grid_layout );
   
   for ( int i = 0; i < 15; i++ )
   {
      QList< QLabel* > row;
      for ( int j = 0; j < 15; j++ )
      {
         QLabel* label;
         label = new QLabel ();
         
         QPixmap zero ( ":/n" );
         label->setPixmap ( zero.scaledToHeight ( 40 , Qt::SmoothTransformation ) );
         
         row.append ( label );
         grid_layout->addWidget ( label , j , i );
      }
      icons.append ( row );
   }
   
   return;
}

void Gui::PuzzleViewer::clean ( void )
{
   for ( int i = 0; i < 15; i++ )
   {
      for ( int j = 0; j < 15; j++ )
      {
         QPixmap zero ( ":/n" );
         icons[ i ][ j ]->setPixmap ( zero.scaledToHeight ( 40 , Qt::SmoothTransformation ) );
         icons[ i ][ j ]->update ();
      }
   }
   
   return;
}

void Gui::PuzzleViewer::update ( Core::Field& field )
{
   for ( int i = 0; i < field.getHeight (); i++ )
   {
      for ( int j = 0; j < field.getWidth (); j++ )
      {
         QPixmap zero ( QString ( ":/%1" ).arg ( field.getPoint ( j , i ) ) );
         icons[ j ][ i ]->setPixmap ( zero.scaledToHeight ( 40 , Qt::SmoothTransformation ) );
         icons[ j ][ i ]->update ();
      }
   }
   
   QApplication::processEvents ();
   
   return;
}

void Gui::PuzzleViewer::setSize ( int new_height , int new_width )
{
   for ( int i = 0; i < new_height; i++ )
   {
      for ( int j = 0; j < new_width; j++ )
      {
         QPixmap zero ( ":/0" );
         icons[ i ][ j ]->setPixmap ( zero.scaledToHeight ( 40 , Qt::SmoothTransformation ) );
         icons[ i ][ j ]->update ();
      }
   }
   
   return;
}
